### ISSUE TITLE
Cache AWS credentials to prevent rate limiting

### ISSUE THREAD

**Issue by nathanielks** [humanmade/hm-platform#128]
AWS is rate limiting authentication requests, so we need to introduce a cache to store our AWS credentials and refresh them when they've expired to prevent fatals.

**Comment by rmccue** [comment]
SC CA seems to be hitting this pretty regularly.

**Comment by nathanielks** [comment]
I've added a ticket internally to prioritize this.

**Comment by joehoyle** [comment]
@nathanielks do you want to code this up, or want me to have a bash at it?

**Comment by nathanielks** [comment]
If you could take it, that'd be great! It'd take me a bit longer to come up with something.

**Comment by joehoyle** [comment]
👍

**Comment by Kenshino** [comment]
This is coming up at least every other day for the CA platform. Takes away focus when we see the bot throw the notification in ``scca-prod-AppStack-4FXSL7MMKD5C-PHPFatalErrorsAlarm-15ZC9YN70RIGJ`: Elevated number of PHP fatal errors. was triggered! Investigate further in Vantage:`

**Comment by nathanielks** [comment]
Looking into this a bit. First place I thought to look was the ElasticPress integration: https://github.com/humanmade/hm-platform/blob/274784c3d0b39c6bc07c98a28fb6c16a53d7c1f3/lib/elasticpress-integration.php#L46-L47

Then I remembered that the `CredentialProvider::defaultProvider()` memoizes credentials, so I don't believe this is it, or there are at least other sources that might be requesting credentials more often as this should save credentials for the whole request.

Then I went to look into S3 Uploads and realized it will do the same by default when no key/secret is passed.

This has me wondering, then. Even with credential memoization, are there too many requests hitting the metadata service or is something else querying it and not memoizing the credentials?

**Comment by nathanielks** [comment]
I figure @joehoyle 	might have more first-hand knowledge before I dive in any further.

**Comment by nathanielks** [comment]
Makes sense. I'd be interested if X-Ray can track requests to the metadata service? That'd be good to visualize just how much is going on.

**Comment by joehoyle** [comment]
Yeha we need to write a middleware for the AWS SDK to send all requests to Xray, I haven't quite found out how to do that. It looks like you can use Guzzle's middleware stack, but the AWS SDK also provides it's own hooking mechanism on requests too.

**Comment by nathanielks** [comment]
@joehoyle I've spent a little time looking into AWS' middleware and playing with the SDK and wanted to report in on what I've found thus far. We have the following information available from a request. We're able to see what URL the request was made to for AWS, but the url itself doesn't show the action was being performed, eg `https://s3.amazonaws.com/`. Are we able to submit any extra information to X-Ray outside of the URL? We're able to read what the action is and the parameters, it's just not in the request itself.

**Comment by joehoyle** [comment]
@nathanielks oh yeah, I had been doing the same! I think I have a very close implementation for this. I was just doin git in some spare time, sorry for not coordingating too. It looks like using a custom `http_handler` to basically what the HTTP handler will work, as that will get the `Command` object which gives you the "high level" description of the request like the AWS service, params etc.

**Comment by joehoyle** [comment]
Also, the AWS SDK is ridiculously obtuse! It's pretty impressive, but soo difficult to pick apart!
